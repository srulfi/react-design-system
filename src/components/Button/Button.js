import React from "react"
import PropTypes from "prop-types"
import classNames from "classnames"
import styled from "styled-components"
import { Button as AntButton } from "antd"

const StyledButton = styled(AntButton)`
	&.quuack-button {
		font-family: ${props => props.theme.fontBase};
		border: 0;
		border-radius: 3em;
		cursor: pointer;
		display: inline-block;
		line-height: 1;
	}
	&.quuack-button--primary {
		background-color: ${props => props.theme.colorPrimary};
		color: ${props => props.theme.textInverseColor};
	}
	&.quuack-button--secondary {
		color: ${props => props.theme.colorSecondary};
		background-color: transparent;
		box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
	}
	&.quuack-button--small {
		font-size: 12px;
		padding: 10px 16px;
	}
	&.quuack-button--medium {
		font-size: 14px;
		padding: 11px 20px;
	}
	&.quuack-button--large {
		font-size: 16px;
		padding: 12px 24px;
	}
`

/**
 * Primary UI component for user interaction
 */
export const Button = ({ primary, backgroundColor, size, label, ...props }) => {
	const mode = primary ? "quuack-button--primary" : "quuack-button--secondary"
	return (
		<StyledButton
			type="button"
			className={classNames(mode, "quuack-button", `quuack-button--${size}`)}
			style={backgroundColor && { backgroundColor }}
			{...props}
		>
			{label}
		</StyledButton>
	)
}

Button.propTypes = {
	/**
	 * Is this the principal call to action on the page?
	 */
	primary: PropTypes.bool,
	/**
	 * What background color to use
	 */
	backgroundColor: PropTypes.string,
	/**
	 * How large should the button be?
	 */
	size: PropTypes.oneOf(["small", "medium", "large"]),
	/**
	 * Button contents
	 */
	label: PropTypes.string.isRequired,
	/**
	 * Optional click handler
	 */
	onClick: PropTypes.func,
}

Button.defaultProps = {
	backgroundColor: null,
	primary: false,
	size: "medium",
	onClick: undefined,
}
