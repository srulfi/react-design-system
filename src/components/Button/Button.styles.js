import styled from "styled-components"
import { Button } from "antd"

import { colors, typography } from "../../shared/styles"

export default styled(Button)`
	&.quuack-button {
		font-family: ${typography.type.fontBase};
		border: 0;
		border-radius: 3em;
		cursor: pointer;
		display: inline-block;
		line-height: 1;
	}
	&.quuack-button--primary {
		background-color: ${colors.primary};
		color: ${colors.white};
	}
	&.quuack-button--secondary {
		color: ${colors.secondary};
		background-color: ${colors.transparent};
		box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
	}
	&.quuack-button--small {
		font-size: ${typography.size.s1}px;
		padding: 10px 16px;
	}
	&.quuack-button--medium {
		font-size: ${typography.size.s2}px;
		padding: 11px 20px;
	}
	&.quuack-button--large {
		font-size: ${typography.size.s3}px;
		padding: 12px 24px;
	}
`
