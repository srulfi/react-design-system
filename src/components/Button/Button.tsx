import React, { FC, ReactNode } from "react"
import classNames from "classnames"

import StyledButton from "./Button.styles"

export interface ButtonProps {
	/**
	 * Is this the principal call to action on the page?
	 */
	primary: boolean
	/**
	 * What background color to use
	 */
	backgroundColor: string
	/**
	 * How large should the button be?
	 */
	size: "small" | "medium" | "large"
	/**
	 * Button content
	 */
	children: ReactNode
	/**
	 * Click handler
	 */
	onClick: () => void
}

/**
 * Primary UI component for user interaction
 */
const Button: FC<ButtonProps> = ({
	primary,
	backgroundColor,
	size,
	children,
	...rest
}) => {
	const mode = primary ? "quuack-button--primary" : "quuack-button--secondary"
	const styles = backgroundColor ? { backgroundColor } : undefined

	return (
		<StyledButton
			className={classNames(mode, "quuack-button", `quuack-button--${size}`)}
			style={styles}
			{...rest}
		>
			{children}
		</StyledButton>
	)
}

Button.defaultProps = {
	backgroundColor: undefined,
	primary: false,
	size: "medium",
}

export default Button
