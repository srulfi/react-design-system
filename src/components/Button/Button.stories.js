import React from "react"

import Button from "./Button"

export default {
	title: "Components/Button",
	component: Button,
	argTypes: {
		backgroundColor: { control: "color" },
	},
}

const Template = args => <Button {...args} />

export const Primary = Template.bind({})
Primary.args = {
	primary: true,
	children: "Button",
	onClick: () => {},
}

export const Secondary = Template.bind({})
Secondary.args = {
	children: "Button",
	onClick: () => {},
}

export const Small = Template.bind({})
Small.args = {
	primary: true,
	size: "small",
	children: "Button",
	onClick: () => {},
}

export const Medium = Template.bind({})
Medium.args = {
	primary: true,
	size: "medium",
	children: "Button",
	onClick: () => {},
}

export const Large = Template.bind({})
Large.args = {
	primary: true,
	size: "large",
	children: "Button",
	onClick: () => {},
}
