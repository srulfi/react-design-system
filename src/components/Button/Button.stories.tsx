import React from "react"
import { Story } from "@storybook/react/types-6-0"

import Button, { ButtonProps } from "./Button"

export default {
	title: "Components/Button",
	component: Button,
	argTypes: {
		backgroundColor: { control: "color" },
	},
}

const Template: Story<ButtonProps> = args => <Button {...args} />

export const Primary = Template.bind({})
Primary.args = {
	primary: true,
	children: "Button",
	onClick: () => console.log("primary button clicked"),
}

export const Secondary = Template.bind({})
Secondary.args = {
	children: "Button",
	onClick: () => console.log("secondary button clicked"),
}

export const Small = Template.bind({})
Small.args = {
	primary: true,
	size: "small",
	children: "Button",
	onClick: () => console.log("primary small button clicked"),
}

export const Medium = Template.bind({})
Medium.args = {
	primary: true,
	size: "medium",
	children: "Button",
	onClick: () => console.log("primary medium button clicked"),
}

export const Large = Template.bind({})
Large.args = {
	primary: true,
	size: "large",
	children: "Button",
	onClick: () => console.log("primary large button clicked"),
}
