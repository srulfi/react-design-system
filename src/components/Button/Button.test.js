import React from "react"
import { shallow } from "enzyme"
import sinon from "sinon"

import Button from "./Button"

describe("test Button Component", () => {
	it("simulates click event", () => {
		const onClick = sinon.spy()
		const button = shallow(<Button onClick={onClick}>Test Button</Button>)
		button.simulate("click")
		expect(onClick).toHaveProperty("callCount", 1)
	})
})
