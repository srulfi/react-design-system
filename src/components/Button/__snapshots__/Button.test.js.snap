// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Button Component renders correctly 1`] = `
<button
  className="ant-btn ant-btn-button sc-bdnylx kIdzGy quuack-button--secondary quuack-button quuack-button--medium"
  onClick={[Function]}
  style={null}
  type="button"
>
  <span>
    Test Button
  </span>
</button>
`;
